doc(kb): mitigation for encrypted RWX volume live expansion (#1145)
longhorn/website: content/kb/troubleshooting-rwx-encrypted-volume-fails-to-live-expand.md

---
title: "Troubleshooting: Encrypted RWX Volume Fails to Perform Live Expansion"
authors:
- "Raphanus Lo"
draft: false
date: 2025-07-01
versions:
- "v1.8.0 to v1.8.1"
- "v1.9.0"
categories:
- "RWX volume"
- "Encrypted volume"
---

## Applicable versions

- v1.8.0 to v1.8.1

> - v1.9.0"

Reviewer: ```suggestion
- v1.9.0
```

## Symptoms

> An encrypted RWX volume created from a PVC and mounted to a workload. While performing the volume expansion, the UI shows the expansion complete without any failure, but the volume keeps to the unresized capacity inside the workload. To the PVC of this volume, it stuck at `FileSystemResizePending` status, and the capacity is outdated:

Reviewer: ```suggestion
An **encrypted RWX volume** created from a PVC and mounted to a workload shows successful expansion in the UI. However, the actual volume inside the workload remains at the original (unexpanded) size.

The corresponding PVC is stuck in the `FileSystemResizePending` phase, with an outdated capacity:
```

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
...
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: "73400320"
  storageClassName: longhorn-crypto-global
  volumeMode: Filesystem
  volumeName: pvc-a01aa079-b05d-40ea-a556-5ba691f9191e
status:
  accessModes:
    - ReadWriteMany
  capacity:
    storage: 50Mi
  conditions:
    - lastProbeTime: null
      lastTransitionTime: "2025-07-01T09:16:05Z"
      message: Waiting for user to (re-)start a pod to finish file system resize of
        volume on node.
      status: "True"
      type: FileSystemResizePending
  phase: Bound
```

> At the meantime, in the corresponding Longhorn share manager logs, it keep complaining "unsupported disk encryption format":

Reviewer: ```suggestion
Meanwhile, the corresponding Longhorn share manager logs repeatedly show errors like:
```

time="2025-07-01T09:31:24Z" level=info msg="Resizing mounted volume" filesystem=/export/pvc-a01aa079-b05d-40ea-a556-5ba691f9191e volume=pvc-a01aa079-b05d-40ea-a556-5ba691f9191e
time="2025-07-01T09:31:24Z" level=info msg="Device /dev/mapper/pvc-a01aa079-b05d-40ea-a556-5ba691f9191e contains filesystem of format ext4" filesystem=/export/pvc-a01aa079-b05d-40ea-a556-5ba691f9191e volume=pvc-a01aa079-b05d-40ea-a556-5ba691f9191e
time="2025-07-01T09:31:24Z" level=error msg="Failed to resize mounted filesystem on volume" error="rpc error: code = InvalidArgument desc = unsupported disk encryption format ext4" filesystem=/export/pvc-a01aa079-b05d-40ea-a556-5ba691f9191e volume=pvc-a01aa079-b05d-40ea-a556-5ba691f9191e
time="2025-07-01T09:31:44Z" level=info msg="Resizing mounted volume" filesystem=/export/pvc-a01aa079-b05d-40ea-a556-5ba691f9191e volume=pvc-a01aa079-b05d-40ea-a556-5ba691f9191e
time="2025-07-01T09:31:44Z" level=info msg="Device /dev/mapper/pvc-a01aa079-b05d-40ea-a556-5ba691f9191e contains filesystem of format ext4" filesystem=/export/pvc-a01aa079-b05d-40ea-a556-5ba691f9191e volume=pvc-a01aa079-b05d-40ea-a556-5ba691f9191e
```

## Root Cause

> During Kubernetes volume live expansion, the CSI plugin node server asks Longhorn share manager to expand the storage capacity. After the capacity gets expanded, the share manager first checks if the underlayer device is a LUKS encrypted device, then resize the mapped device to sync the capacity. But the share manager actually checks the type of mapped device, hence CSI plugin node server keeps receiving resize failure and consider the expansion is in complete.

Reviewer: ```suggestion
During live volume expansion in Kubernetes, the CSI plugin's node server instructs the Longhorn share manager to resize the volume. The share manager checks whether the underlying device is LUKS-encrypted before resizing.

However, instead of validating LUKS encryption properly, it mistakenly checks the mapped device's filesystem type. This leads to a failure in the resize operation, causing the PVC to remain in a pending state, even though the expansion appears complete in the UI.
```

Reviewer 2: > the share manager actually checks the type of mapped device

This statement is confusing for regular users. They may not know that `mapped device` means encrypted device path. Besides, what error does it cause when Longhorn uses a regular mounter to check disk format of the encrypted device path?

Author: The share manager internally checks two point before resizing the `/dev/mapper/...` device:

1. It expect the `/dev/mapper/<volume>` to be a LUKS device. The correct LUKS device is `/dev/longhorn/<volume>` and be encrypted as `/dev/mapper/<volume>` ext4 disk. This is what the user currently meets in share manager's log.
    > time="2025-07-01T09:31:44Z" level=info msg="Device /dev/mapper/pvc-a01aa079-b05d-40ea-a556-5ba691f9191e contains filesystem of format ext4" filesystem=/export/pvc-a01aa079-b05d-40ea-a556-5ba691f9191e volume=pvc-a01aa079-b05d-40ea-a556-5ba691f9191e
1. After passing the format check, it will also check the status of encrypted volume. [Share manager cannot properly parse the status from "/dev/mapper/<volume> is active and is in use."](https://github.com/longhorn/longhorn-share-manager/blob/v1.9.0/pkg/crypto/crypto.go#L118)

This issue is not reflected on the UI. Without the checking the logs, the user will only notice this issue when the storage reaches its maximum capacity, or by checking the disk capacity inside their workload.

Reviewer 2: In this statement, it's better to say something like: 
```
Theoretically, Longhorn should check the disk format of the original device and get the result "crypto_LUKS". However, it wrongly fetches that of the encrypted/mapped device with the result `ext4`. Then Longhorn can not resize the encrypted device correctly.
```

## Mitigation

> ### Option A. Scale Down The Workload And Fix The State

Reviewer 3: Any potential side effect?

Author: After discussion, we prefer not to touch the PVC. Will keep the only way to help user upgrading Longhorn.

1. Scale the workload replica to 0, and wait for the volume gets detached.

> 2. Edit the corresponding PV, make sure the volume reclaim policy is `Retain`, and remove the `claimRef`
> ```yaml
> apiVersion: v1
> kind: PersistentVolume
> ...
> spec:
>   accessModes:
>     - ReadWriteMany
>   persistentVolumeReclaimPolicy: Retain
> ```

Reviewer: ```suggestion
2. Edit the corresponding PV:
  - Ensure `persistentVolumeReclaimPolicy` is set to `Retain`.
  - Remove the `claimRef` field.
    ```yaml
    apiVersion: v1
    kind: PersistentVolume
    ...
    spec:
      accessModes:
        - ReadWriteMany
      persistentVolumeReclaimPolicy: Retain
    ```
```

3. `kubectl delete pvc <pvc_name>`

> 4. In the `Volume` page of the Longhorn UI, choose the workload's RWX volume, click `create PV/PVC` in the operation menu, check `Create PVC` and `Use Previous PVC`, then click `OK` to recreate the PVC.

Reviewer: ```suggestion
4. In the Longhorn UI > Volume page, locate the affected RWX volume.
  - Click "Create PV/PVC" from the operation menu.
  - Select "Create PVC" and "Use Previous PVC", then click OK.
```

> 5. Confirm the PVC creation in Kubernetes. The capacity should be corrected.
> 6. Scale the workload back. The expanded volume will be attached.
> 7. Edit the PV's reclaim policy back if needed.

Reviewer: ```suggestion
5. Confirm that the PVC is recreated in Kubernetes and shows the correct capacity.
6. Scale the workload back up, the expanded volume will be attached.
7. (Optional) Revert the PV's reclaim policy if needed.
```

### Option B. Upgrade To Fixed Longhorn Version

This method does not shut the workload pods down, but will introduce an I/O gap for share manager pod recreation.

1. Upgrade to Longhorn v1.10.0 or newer.

> 2. `kubectl -n longhorn-system edit sharemanager <volume_name>`, and upgrade the image version in the spec.

Reviewer: ```suggestion
2. `kubectl -n longhorn-system edit sharemanager <volume_name>`, and upgrade the image version in the spec.
```
```suggestion
2. Edit the Share Manager to update the image:
    ```
    kubectl -n longhorn-system edit sharemanager <volume_name>
    ```
    In the spec, update:
```

    ```
    apiVersion: longhorn.io/v1beta2
    kind: ShareManager
    ...
    spec:
      image: longhornio/longhorn-share-manager:v1.10.0
    ...
    ```

> 3. `kubectl -n longhorn-system edit pod share-manager-<volume_name>`, and upgrade the image version in the spec.

Reviewer: ```suggestion
3. Edit the Share Manager pod:
    ```
    kubectl -n longhorn-system edit pod share-manager-<volume_name>
    ```
    Update the container image:
```

    ```
    apiVersion: v1
    kind: Pod
    ...
    spec:
      containers:
      - image: longhornio/longhorn-share-manager:v1.10.0
    ...
    ```

> 4. The volume I/O in the workload will be stuck for a while, depends on the time to recreate the share manager pod.

Reviewer: ```suggestion
4. The volume I/O in the workload will be stuck for a while as the share manager pod restarts.
```
